### PR TITLE
[BC-breaking] Introduced InterpolationModes and deprecated arguments: resample and fillcolor

### PR DESCRIPTION
**Pull request by vfdev-5**
Related to #2479

- Introduced `InterpolationModes`
- Deprecated arguments: `resample` and `fillcolor`
- Replaced by `interpolation` and `fill`

Context: 
1) `fill` is mainly used: 4 times in transforms and 3 times in functional vs `fillcolor` is used 2 times in transforms and functional (RandomAffine and F.affine) => we can decide to replace `fillcolor` by `fill` and set a warning about argument deprecation.

2) `interpolation` is used 6 times in transforms/functional vs `resample` is used 4 times in transforms/functional => we can similarly decide to replace `resample` by `interpolation` and set a warning about argument deprecation.

### CODE REVIEW

File: torchvision/transforms/functional.py
Comment (datumbox) [inline torchvision/transforms/functional.py:829]: Should we be honouring the provided `resample` value if `interpolation` is not provided?

File: torchvision/transforms/functional.py
Comment (datumbox) [inline torchvision/transforms/functional.py:886]: Same here.

File: torchvision/transforms/transforms.py
Comment (datumbox) [inline torchvision/transforms/transforms.py:1160]: Same here.

File: torchvision/transforms/transforms.py
Comment (datumbox) [inline torchvision/transforms/transforms.py:1249]: same here.

**Review by datumbox — COMMENTED** [review]
@vfdev-5 Thanks for the PR.

I added one suggestion to improve BC during the deprecation period and marked all the places where this could happen. Nevertheless, if our policy is typically to only throw warnings feel free to ignore.

Also I noticed that in some methods you throw warnings while in others you directly replace the names. I assume that this is because the first are considered public API while the latter not. If that's true, then all good.

File: torchvision/transforms/functional.py
Comment (vfdev-5) [reply torchvision/transforms/functional.py:829]: Do you mean, set `interpolation=resample` if `resample is not None` ? True that it would lead to an unexpected result if called like
```
F.rotate(x, 45, resample=2) == F.rotate(x, 45, interpolation=0)
```

File: torchvision/transforms/transforms.py
Comment (vfdev-5) [reply torchvision/transforms/transforms.py:1160]: Right, here as well we can set `interpolation=resample`

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the review @datumbox 

> I added one suggestion to improve BC during the deprecation period and marked all the places where this could happen. Nevertheless, if our policy is typically to only throw warnings feel free to ignore.

I agree that it is a good practice to provide until when the argument will remain in the API. I'd say by default, it can be 2 future minor releases, until 0.10.0. @fmassa any thoughts?

File: torchvision/transforms/functional.py
Comment (datumbox) [reply torchvision/transforms/functional.py:829]: Yes exactly. :)

**Review by datumbox — APPROVED** [review]
Thanks for the updates @vfdev-5, your last changes look great.

@fmassa Are you OK to merge?

File: torchvision/transforms/functional.py
Comment (fmassa) [inline torchvision/transforms/functional.py:890]: nit: I'm not a native English speaker, but I feel like `since` might not be the word here. I would have said instead
```suggestion
            "Argument resample is deprecated and will be removed in v0.10.0. Please, use interpolation instead"
```

This is a minor nit (which applies to the other messages as well), so I'm not considering this as a merge-blocking issue.

**Review by fmassa — COMMENTED** [review]
Thanks for the PR @vfdev-5 !

I have one comment regarding the warning message, but it is not blocking for merge.

On a separate note that I would like to discuss here:

When was planning for the Tensor support for transforms, I started to wonder if we should move away from the `interpolate=0` enums from PIL (= `Image.NEAREST` etc), and instead use a more descriptive argument which could work even without the knowledge of PIL.

Two options came to my mind:
- replace `interpolate : int` with `mode : str` as we have in `torch.nn.functional.interpolate`
- provide our own Enum for the interpolation mode in torchvision (and hope it works on torchscript)

Thoughts?

**Comment by vfdev-5** [comment]
If we would like to use an enum for interpolate, there can be two potential problems with torchscript:
```python
import torch
from enum import Enum

class Modes:
    NEAREST = 0
    BILINEAR = 1

def func(x: torch.Tensor, mode: Modes) -> torch.Tensor:
    if mode == Modes.NEAREST:
        return x + 1
    elif mode == Modes.BILINEAR:
        return x + 2
    else:
        return x

    
y = func(torch.rand(3, 4), Modes.NEAREST)
s_func = torch.jit.script(func)
``` 

1) if this code is launched inside jupyter gives an error:
```
/opt/conda/lib/python3.7/site-packages/torch/_utils_internal.py in get_source_lines_and_file(obj, error_msg)
     46     filename = None  # in case getsourcefile throws
     47     try:
---> 48         filename = inspect.getsourcefile(obj)
     49         sourcelines, file_lineno = inspect.getsourcelines(obj)
     50     except OSError as e:

/opt/conda/lib/python3.7/inspect.py in getsourcefile(object)
    682     Return None if no way can be identified to get the source.
    683     """
--> 684     filename = getfile(object)
    685     all_bytecode_suffixes = importlib.machinery.DEBUG_BYTECODE_SUFFIXES[:]
    686     all_bytecode_suffixes += importlib.machinery.OPTIMIZED_BYTECODE_SUFFIXES[:]

/opt/conda/lib/python3.7/inspect.py in getfile(object)
    651             if getattr(module, '__file__', None):
    652                 return module.__file__
--> 653         raise TypeError('{!r} is a built-in class'.format(object))
    654     if ismethod(object):
    655         object = object.__func__

TypeError: <class '__main__.Modes'> is a built-in class
```

2) if this code is launched as a script with `if __name__ == "__main__":` close, it gives another error:
```
Traceback (most recent call last):
  File "torchscript_enum.py", line 22, in <module>
    s_func = torch.jit.script(func)
  File "/opt/conda/lib/python3.7/site-packages/torch/jit/_script.py", line 940, in script
    qualified_name, ast, _rcb, get_default_args(obj)
RuntimeError: 
python value of type 'str' cannot be used as a value. Perhaps it is a closed over global variable? If so, please consider passing it in as an argument or use a local varible instead.:
  File "torchscript_enum.py", line 11
def func(x: torch.Tensor, mode: Modes) -> torch.Tensor:
    if mode == Modes.NEAREST:
               ~~~~~~~~~~~~~ <--- HERE
        return x + 1
    elif mode == Modes.BILINEAR:
```

**Comment by vfdev-5** [comment]
However, the following works inside the script file 
```python
import torch
from enum import Enum

global Modes

class Modes(Enum):
    NEAREST = 0
    BILINEAR = 1


def func(x: torch.Tensor, mode: Modes) -> torch.Tensor:
    if mode == Modes.NEAREST:
        return x + 1
    elif mode == Modes.BILINEAR:
        return x + 2
    else:
        return x

if __name__ == "__main__":
    y = func(torch.rand(3, 4), Modes.NEAREST)
    s_func = torch.jit.script(func)
    y = s_func(torch.rand(3, 4), Modes.NEAREST)
```

According to https://github.com/pytorch/pytorch/pull/42963

**Comment by fmassa** [comment]
It's great that we can use Enum's with torchscript, but it's weird that we would need to have it set as a `global` beforehand.
cc @gmagogsfm is this expected?

**Comment by vfdev-5** [comment]
Sorry, seems like `global` keyword can be omitted if used inside a script file.

File: torchvision/transforms/functional_pil.py
Comment (datumbox) [inline torchvision/transforms/functional_pil.py:477]: I assume that JIT will complain if you change the numeric default value with the enum, correct?

Comment (vfdev-5) [reply torchvision/transforms/functional_pil.py:477]: This is private PIL affine method and in addition torchscript should not pass here.

File: torchvision/transforms/functional_tensor.py
Comment (datumbox) [inline torchvision/transforms/functional_tensor.py:760]: Again here, why we use the string value instead of its equivalent enum?

Comment (vfdev-5) [reply torchvision/transforms/functional_tensor.py:760]: Because, such that we should not map it to the arguments for torch functions.

File: torchvision/transforms/functional.py
Comment (datumbox) [inline torchvision/transforms/functional.py:288]: Just marking this as a BC-breaking change.

**Review by datumbox — APPROVED** [review]
The PR looks good. Thanks a lot for the changes, they look great.

I understand that it's a design decision to maintain BC on the high-level transforms.py but not with the lower-level functional implementations, correct? Also, I noticed that on the backend implementations you use raw values for default (ints or strings) instead of their enums. Is this because of a JIT limitation?

**Comment by vfdev-5** [comment]
> I understand that it's a design decision to maintain BC on the high-level transforms.py but not with the lower-level functional implementations, correct? 

BC change is "required" to switch from PIL.Image resampling mode. No change in lower-level functional implementations is a sort of minimal change design. 

> Also, I noticed that on the backend implementations you use raw values for default (ints or strings) instead of their enums. Is this because of a JIT limitation?

We can certainly find a way to reuse `InterpolationModes` inside private modules. As I said this is currently blocked by cyclic imports due to where `InterpolationModes` is defined. Remap from InterpolationModes to PIL modes happens inside `F` for each method. Remap from InterpolationModes to string values for torch functions is done with `InterpolationModes.value` attribute.

File: torchvision/transforms/functional_pil.py
Comment (datumbox) [reply torchvision/transforms/functional_pil.py:477]: I suppose we could move the definitions to avoid circular dependencies but I understand that this PR has been dragging for weeks now, so I don't think we should block it further for this.

Comment (vfdev-5) [reply torchvision/transforms/functional_pil.py:477]: IMO, the most suitable option is to create another module with modes, such that `F`, `F_t` and `F_pil` could import it and reuse if needed. If there is an argumented need to do that, I'm happy to recoded no matter of how many time the PR is open... IMO, it is better to properly design things and go slower that faster and have a poor API :)

Comment (datumbox) [reply torchvision/transforms/functional_pil.py:477]: I think your proposed changes on the transforms API are very elegant. Your choice of using enums instead of raw values has multiple benefits (less mistakes, clearer API, easier to search usages etc), so as a user I would definitely love it to be similar across all functional APIs. Again I would not say this is a deal breaker for me but more of a nice-to-have.

File: torchvision/transforms/functional.py
Comment (fmassa) [inline torchvision/transforms/functional.py:332]: This makes for a big BC-breaking change, as all other code that users had before won't work anymore because of this.

I think it might be worth guarding against `interpolation` being an `int`, something like
```python
if not isinstance(interpolation, InterpolationModes):
    # wrap in an Enum
```
otherwise IIUC all code from before won't work anymore, which I think is not something we can do.

**Review by fmassa — CHANGES_REQUESTED** [review]
Correct me if I'm wrong, but I think the BC-breaking change that is introduced in here will break all user code that uses the transforms and I if that's the case I'm not sure this is something we can do in torchvision.

File: torchvision/transforms/functional.py
Comment (vfdev-5) [reply torchvision/transforms/functional.py:332]: Yes, makes senses. Let me see if we can do something with that.

File: torchvision/transforms/functional.py
Comment (fmassa) [inline torchvision/transforms/functional.py:909]:
```suggestion
            "Argument resample is deprecated and will be removed in v0.10.0. Please, use interpolation instead"
```

**Review by fmassa — APPROVED** [review]
Thanks a lot!

I'm merging the PR now to unblock, but I have one last comment that I would like us to check: should it be `InterpolationMode` or `InterpolationModes`?

cc @vfdev-5 @datumbox for thoughts

**Comment by datumbox** [comment]
@fmassa I think enums are typically singular. Aka `Color.Blue` not `Colors.Blue`. 

On a similar enum, I used `ImageReadMode`.

**Comment by vfdev-5** [comment]
@fmassa @datumbox thanks for pointing out that. I agree that I can rename `InterpolationModes` to `InterpolationMode`.

**Pull request merged**
